### ISSUE TITLE
TreeNode accessibility improvements

### ISSUE THREAD

**Issue by Garbee** [devtools-html/devtools-core#104]
When running an [a11y audit](https://github.com/GoogleChrome/accessibility-developer-tools) on the debugger it highlights the file manager in particular (ignoring the syntax highlighting/codemirror stuff.)

This appears to come from the [button](https://github.com/devtools-html/devtools-core/blob/dd29b3f99fc8301f3861e2a7d20a58fcab9a73a0/packages/devtools-sham-modules/client/shared/components/tree.js#L99) used within the tree nodes to make them focusable on MacOS.

There are two things we could do here to address this:

1. Add a label for the button so it gets read aloud. This however would mean we need to silence the upper-level so the name isn't read twice by screen readers.
2. We could refactor the markup so that the button is the actual node container. This means more CSS to override the button defaults, but gains us fewer DOM nodes which is important to minimize with larger applications being created that need to be debugged.

![selection_020](https://cloud.githubusercontent.com/assets/868301/22153236/76e45534-def4-11e6-97d6-7b8a6da3cbe1.png)
 
![selection_019](https://cloud.githubusercontent.com/assets/868301/22153240/7b236b80-def4-11e6-93fc-c1308278718a.png)

**Comment by Garbee** [comment]
Another solution to this would be to refactor the tree view in whole. Make it so if a node has children, it renders using something like the `details` element that naturally handles the children display. Which we'd continue to lazily load and render upon open/close. But if the item has no children (is a leaf at the end of the tree) then it would be a direct button to toggle the file opening in the source view.

This would alleviate quite a bit of complexity for the tree view by putting the accessibility pieces back into the hands of the browser directly for the actions. All we would be responsible for is re-styling the default look. The downside to this approach however, is no [Edge/IE support](http://caniuse.com/#feat=details) for `<details>` if anyone were running the debugger's frontend with that engine.

**Comment by nchevobbe** [comment]
> Another solution to this would be to refactor the tree view in whole. Make it so if a node has children, it renders using something like the details element that naturally handles the children display. Which we'd continue to lazily load and render upon open/close. But if the item has no children (is a leaf at the end of the tree) then it would be a direct button to toggle the file opening in the source view.

i quite like this idea :)
One other downside though is that styling the "▶" isn't easy to do. There is an non-standard `summary::-webkit-details-marker` but nothing alike in Firefox. And even in Chrome, if we would have to do a simple animation, as we do now, it wouldn't be as simple I think.
But I guess this is something worth digging

**Comment by Garbee** [comment]
Yup, the built-in arrow is the major holdup with this if we wanted to restyle the details summary view. Worth bringing that up though in the WICG to discuss standardizing access to that.

**Comment by Garbee** [comment]
Just testing with Chromevox now against it, it is very bad for screen readers. All that you get tabbing through the file list is `button` read out. No context as to what the button is for.

**Comment by Garbee** [comment]
I've brought up the styling issue on the [WICG Discourse](https://discourse.wicg.io/t/ability-to-style-summary-element-arrow/1969). This kind of thing won't be quick to get done though. I'm starting research through for recommendations for handling on this type of interaction without details.

**Comment by nchevobbe** [comment]
If I remember correctly, @SebastianZ has some experience with specs stuff.
if so @SebastianZ, any hint for @Garbee ?

**Comment by Garbee** [comment]
I think CSS's visibility state will actually handle a lot of the a11y we need automatically. That is all a `details` element is doing. What we need to do though primarily is make the headers buttons themselves and use the `::before` pseudo to provide the arrow. If we wanted to provide a more explicit label, then we could add `aria-label` to the buttons as well.

<details>
<summary>Example Markup</summary>

```html
<div class="accordion" data-open="false">
  <button>domain.com</button>
  <div class="accordion__content">
    <div class="accordion" data-open="false">
        <button>somepath</button>
        <div class="accordion__content">
            <div>
               <label><input type=checkbox checked>todo-view.js</label>
            </div>
            <div>
               <label><input type=checkbox>app-view.js</label>
            </div>
        </div>
    </div>
  </div>
</div>
```

</details>

In this example our accordion is a div container. It has buttons as children which are our toggles for sections, or the final input toggles to open files in the file manager. These are checkbox inputs (that would be `0 x 0` size so they don't render) so we get the benefit of the native reading of "checked" if a file is already open in the file list. Overall, this structure markup is superior and we only toggle visibility of the next child of an accordion header on click.

All the same thing `<details>` provides right now without any of the hindrances of its styling.

**Comment by Garbee** [comment]
smaug in the Discourse discussion brought up the WhatWG spec. The rendering should be that of `list-style` which let's use do `list-style: none` to remove the built in arrow. Firefox adheres to this, so that is all that is needed here. We can then use the pseudo selector for Blink/webkit until they change to adhere to the spec. I have pinged for an update on the current [Blink bug](https://bugs.chromium.org/p/chromium/issues/detail?id=590014) related to this change.

**Comment by SebastianZ** [comment]
> If I remember correctly, @SebastianZ has some experience with specs stuff.
> if so @SebastianZ, any hint for @Garbee ?

Here are some related links:
* [Definition of the `::marker` pseudo-element](https://drafts.csswg.org/css-lists/#marker-pseudo-element) allowing to style the triangle
* [Bug 1221416](https://bugzilla.mozilla.org/show_bug.cgi?id=1221416) for the implementation of the `::marker` pseudo-element
* [Discussion about `<summary>` and `::marker` on www-style](https://lists.w3.org/Archives/Public/www-style/2016Apr/0365.html)
* [Request to add an `:open` pseudo-class for `<details>`](https://github.com/w3c/csswg-drafts/issues/611)

As @Garbee wrote, the best solution for now is to use `list-style: none` on the `<summary>` element and add the marker via the `::before` pseudo-element.

Sebastian

**Comment by nchevobbe** [comment]
Thanks @SebastianZ !

**Comment by Garbee** [comment]
[Codepen](http://codepen.io/Garbee/pen/jywdZr) showing what is workable in FF, Safari (as it uses the pseudo element as well looking at their layout tests), and Chrome. I think this is the best path forward for refactoring the markup so we aren't managing the keyboard events to toggle the display. We just let the browser do that lifting. The arrow can either be inserted with `::before` or we can just put the arrow into the markup like we currently do and verify it is hidden from screen readers.

I'll be working on implementing this behind a feature flag today so we can have a proof-of-concept to play with.

**Comment by Garbee** [comment]
So, [example codepen](http://codepen.io/Garbee/pen/pRrjWa) showing my initial happy point. If I can get this to render with our lazy-loading subtrees then I think we'll have a good start to make further modifications from and start getting developer feedback from behind a feature flag.

(Note: The material icons is just for the demo arrow, we'll still be using the SVG internally.)
